ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+# Ruby SDK - Phase 1
+
+Ruby is a popular dynamic programming language that many companies are leaning on to beat their
+competitors to the market. With the focus on the speed of execution it is very common for these
+companies to rely on managed solutions. This presents an opportunity for Temporal to become Ruby's
+de facto standard for workflow orchestration.
+
+
+## Timeline
+
+The Ruby SDK will be implemented in four loosely defined phases:
+
+- Phase 1 — initial setup, activity APIs, client and worker, payload converters
+- Phase 2 — per-class execution attributes, logging, metrics, error handling, middleware
+- Phase 3 — workflow APIs, local activities, async activity completion, etc
+- Phase 4 — Sandbox, replayer, support for testing
+
+
+## Overview
+
+The Ruby SDK will be implemented on top of the [SDK Core](https://github.com/temporalio/sdk-core).
+The exact bridging mechanism will be decided later as there are different alternatives.
+
+### Goals
+
+- Quick path to a full working setup
+- A path of minimal divergence from other SDKs (does allow for other approaches, but it's not the
+  goal)
+- Minimise external dependencies (gems)
+- Local and end-to-end testability of workflows
+- Type safety where possible (optional and exact approach to be chosen later)
+- Compatibility with existing tools (error handling, logging, metrics, middleware, etc)
+
+### Non-goals
+
+- Convention over configuration (in favour of cross-SDK interface consistency, this goes against
+  Ruby principles)
+- Backwards compatibility with Coinbase's Ruby SDK (current SDK customers will have to do a full
+  migration)
+- Reducing the number of different approaches to achieving the same goal (let the SDK user choose
+  what's best for their application/taste)
+
+### Tech stack && approach
+
+- Ruby version >= 2.7 (currently oldest fully supported version)
+- Type safety based on RBS or Sorbet (RBS will require Ruby version => 3.0)
```

Comment:
Great! Sounds like RBS is where it's at. A quick glance even shows recent support for bounded generics and other neat things.

So my suggestion here is that we only keep the client `start_workflow`/`execute_workflow`/`signal`/`query` invocations using strings in the signatures until we know more about workflow API definitions in later phases.

In the phase, here's the challenge: Can you show me what a `start_workflow` type signature would look like that, when I pass a class reference (or workflow "run"/"execute" proc) that accepts an integer argument, it will fail the type checker when I call it with a string argument in `start_workflow`? I know we don't necessarily want types to dictate our design, but I could see this making you want to use unbound methods like `client.start_workflow(MyWorkflow.run,'some param')` to get benefits of generics if that's possible.

Reply:
Great point 👍 I'll dig deeper into this and put together a prototype of what is possible and what isn't